src/components/AttachmentCarousel/index.js

```diff
@@ -0,0 +1,229 @@
+
+    /** Callback to update the parent modal's state with a sourceUrl and name from the attachments array */
+    onNavigate: PropTypes.func,
+
+    /** The report currently being looked at */
+    report: reportPropTypes,
+
+    /** Object of report actions for this report */
+    reportActions: PropTypes.objectOf(PropTypes.shape(reportActionPropTypes)),
+
+};
+
+const defaultProps = {
+    sourceURL: '',
+    report: {
+        isLoadingMoreReportActions: false,
+        reportID: '',
+    },
+    reportActions: {},
+    onNavigate: () => {},
+};
+
+class AttachmentCarousel extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.canUseTouchScreen = canUseTouchScreen();
+        this.makeStateWithReports = this.makeStateWithReports.bind(this);
+        this.cycleThroughAttachments = this.cycleThroughAttachments.bind(this);
+        this.onShowArrow = this.onShowArrow.bind(this);
+
+        this.state = {shouldShowArrow: this.canUseTouchScreen};
+    }
+
+    componentDidMount() {
+        this.makeStateWithReports();
+    }
+
+    componentDidUpdate(prevProps) {
+        const prevReportSize = _.size(prevProps.reportActions);
+        const currReportSize = _.size(this.props.reportActions);
+        if (prevReportSize === currReportSize) {
+            return;
+        }
+        this.makeStateWithReports();
+    }
+
+    /**
+     * Toggles the visibility of the arrows
+     * @param {Boolean} shouldShowArrow
+     */
+    onShowArrow(shouldShowArrow) {
+        this.setState({shouldShowArrow});
+    }
+
+    /**
+     * Helps to navigate between next/previous attachments
+     * @param {Object} attachmentItem
+     * @returns {Object}
+     */
+    getAttachment(attachmentItem) {
+        const sourceURL = _.get(attachmentItem, 'sourceURL', '');
+        const file = _.get(attachmentItem, 'file', {name: ''});
+        this.props.onNavigate({sourceURL: addEncryptedAuthTokenToURL(sourceURL), file});
+
+        return {
+            sourceURL,
+            file,
+        };
+    }
+
+    /**
+     * This is called when there are new reports to set the state
+     */
+    makeStateWithReports() {
+        let page;
+        const actions = ReportActionsUtils.getSortedReportActions(_.values(this.props.reportActions), true);
+
+        /**
+         * Calling reducer will filter out attachments,
+         * determine the index of opened attachment,
+         * and retrieve the src url and name of attachements
+         */
+        const attachments = [];
+        _.forEach(actions, ({originalMessage}) => {
+            if (originalMessage && originalMessage.html) {
+                const matchesIt = originalMessage.html.matchAll(CONST.REGEX.ATTACHMENT_DATA);
+                const matches = [...matchesIt];
+
+                // matchAll captured both source url and name of the attachment
+                if (matches.length === 2) {
+                    const [sourceURL, name] = _.map(matches, m => m[2]);
```

Comment:
@JediWattson, as a heads up, carlos comment  mentioned [here](https://github.com/Expensify/App/pull/9279#issuecomment-1352390422) has not been tackled.

The sourceURL in actions is not necessarily the ideal source of truth. This is because for people working using ngrok (which is a lot of expensify peeps) we replace the image URL with Config.EXPENSIFY.URL_API_ROOT as you can see [here](https://github.com/Expensify/App/blob/3439796754d31accfdd4b116afe0eba598278f1b/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L42-L45). This is how people still see their images using ngrok. You would need to find a way to use the URL using URL_API_ROOT. Maybe calling,
```
.replace(
    Config.EXPENSIFY.EXPENSIFY_URL,
    Config.EXPENSIFY.URL_API_ROOT,
);
```
before you push the sourceURL would work.

Reply:
@chiragsalian I think I got it [here!](https://github.com/Expensify/App/pull/9279/commits/7b7a5a0a9f990a89b830751a6d4062f5f2038c37#diff-50277d0d01df2562c4e370343d4616a461f4dbdcbf0b94c151c720659c320016R89)

Reply:
i wasn't able to test today. I'll test it tomorrow morning.

Reply:
Tested and it did not work. This might help, when we reach the AttachmentCarousel index.js component,
- An example this.props.sourceURL would be `"https://expensify-chirag.ngrok.io/chat-attachments/5757/w_c6870736eeea9ef20dba424599ff95e4555df844.jpg"`
- But the map [here](https://github.com/Expensify/App/blob/7b7a5a0a9f990a89b830751a6d4062f5f2038c37/src/components/AttachmentCarousel/index.js#L121) would set sourceURL to `"https://www.expensify.com.dev/chat-attachments/5757/w_c6870736eeea9ef20dba424599ff95e4555df844.jpg"`
- So then few lines down over [here](https://github.com/Expensify/App/blob/7b7a5a0a9f990a89b830751a6d4062f5f2038c37/src/components/AttachmentCarousel/index.js#L122) the `sourceURL.includes(this.props.sourceURL)` check will fail. 

So page will be set and then `this.getAttachment(attachments[page])` passes undefined to getAttachment and hence nothing is returned.

The problem should be solved by doing the following in `makeStateWithReports`,
```jsx
let [sourceURL, name] = _.map(matches, m => m[2]);

// Update the image URL so the images can be accessed depending on the config environment.
// Eg: while using Ngrok the image path is from an Ngrok URL and not an Expensify URL.
sourceURL = sourceURL.replace(
    Config.EXPENSIFY.EXPENSIFY_URL,
    Config.EXPENSIFY.URL_API_ROOT,
);
```
and then reverting the change you made [here](https://github.com/Expensify/App/pull/9279/commits/7b7a5a0a9f990a89b830751a6d4062f5f2038c37#diff-50277d0d01df2562c4e370343d4616a461f4dbdcbf0b94c151c720659c320016R89).